libr/anal/var.c

```diff
@@ -136,272 +159,115 @@ R_API bool r_anal_var_add(RAnal *a, ut64 addr, int scope, int delta, char kind,
 		break;
 	default:
 		eprintf ("Invalid var kind '%c'\n", kind);
-		return false;
+		return NULL;
 	}
 	if (kind == R_ANAL_VAR_KIND_REG) {
-		reg = r_reg_index_get (a->reg, R_ABS (delta));
+		reg = r_reg_index_get (fcn->anal->reg, R_ABS (delta));
 		if (!reg) {
 			eprintf ("Register wasn't found at the given delta\n");
-			return false;
+			return NULL;
 		}
 	}
-	const char *var_def = sdb_fmt ("%d,%s,%d,%s,%s", isarg, type, size, name, reg? reg->name : NULL);
-	if (scope > 0) {
-		const char *sign = "";
-		if (delta < 0) {
-			delta = -delta;
-			sign = "_";
-		}
-		/* local variable */
-		const char *fcn_key = sdb_fmt ("fcn.0x%"PFMT64x ".%c", addr, kind);
-		const char *var_key = sdb_fmt ("var.0x%"PFMT64x ".%c.%d.%s%d", addr, kind, scope, sign, delta);
-		const char *name_key = sdb_fmt ("var.0x%"PFMT64x ".%d.%s", addr, scope, name);
-		const char *shortvar = sdb_fmt ("%d.%s%d", scope, sign, delta);
-		sdb_array_add (DB, fcn_key, shortvar, 0);
-		sdb_set (DB, var_key, var_def, 0);
-		if (*sign) {
-			delta = -delta;
+	RAnalVar *var = r_anal_function_get_var (fcn, kind, delta);
+	if (!var) {
+		var = R_NEW0 (RAnalVar);
+		if (!var) {
+			return NULL;
 		}
-		char *name_val = r_str_newf ("%c,%d", kind, delta);
-		sdb_set (DB, name_key, name_val, 0);
-		free (name_val);
-	} else {
-		/* global variable */
-		const char *var_global = sdb_fmt ("var.0x%"PFMT64x, addr);
-		const char *var_def = sdb_fmt ("%c.%s,%d,%s", kind, type, size, name);
-		sdb_array_add (DB, var_global, var_def, 0);
-	}
-	return true;
+		r_pvector_push (&fcn->vars, var);
+	}
+	var->name = strdup (name);
+	var->regname = reg ? strdup (reg->name) : NULL; // TODO: no strdup here? pool? or not keep regname at all?
+	var->type = strdup (type);
```

Comment:
maybe one day we can have refcounted strings and reduce the memory uasge of r2 in half